pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"io"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+type PodSandboxID string
+
+// PodSandboxManager provides basic operations to create/delete and examine the
+// PodSandboxes. These methods should either return an error or block until the
+// operation succeeds.
+type PodSandboxManager interface {
+	// Create creates a sandbox based on the given config, and returns the
+	//  of the new sandbox.
+	Create(config *PodSandboxConfig) (PodSandboxID, error)
+	// Delete deletes the sandbox by its ID. If there are any running
+	// containers in the sandbox, they will be terminated as a side-effect.
+	Delete(id PodSandboxID) error
+	// List lists existing sandboxes, filtered by the given PodSandboxFilter.
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	// Status gets the status of the sandbox by ID.
+	Status(id PodSandboxID) (PodSandboxStatus, error)
+	// PortForward copies the data from the port in the PodSandBox to the
+	// stream.
+	// TODO: Should PortForward be in a separate interface.
+	PortForward(id PodSandboxID, port uint16, stream io.ReadWriteCloser) error
+}
+
+// PodSandboxConfig holds all the required and optional fields for creating a
+// sandbox.
+type PodSandboxConfig struct {
+	// Name is the name of the sandbox. The string should conform to
+	// [a-zA-Z0-9_-]+.
+	Name string
+	// Hostname is the hostname of the sandbox.
+	Hostname string
+	// DNSOptions sets the DNS options for the sandbox.
+	DNSOptions DNSOptions
+	// PortMappings lists the port mappings for the sandbox.
+	PortMappings []PortMapping
+	// Resources specifies the resource requirements for the sandbox.
+	// Note: On a Linux host, kubelet will create a pod-level cgroup and pass
+	// it as the cgroup parent for the PodSandbox. For some runtimes, this is
+	// sufficent to fulfill pod-level resource requirements, and this field
+	// will not be used. For others, e.g., hypervisor-based runtimes, explicit
+	// resource requirements for the sandbox are needed at creation time.
+	Resources PodSandboxResources
+	// Path to the directory on the host in which container log files are
+	// stored.
+	LogDirectory string
+	// Labels are key value pairs that may be used to scope and select
+	// individual resources.
+	Labels Labels
```

Comment:
The selectors will have a use; they'll tell the runtime that they must provide a way to list containers by PodUID,etc. The fact that that's done with labels in the docker runtime is an implementation detail which I don't think necessarily has to be exposed in this interface. E.g. the docker manager can create those labels even though they're not specified via the create call because it knows that to support `List(PodUID=foo)` it will filter by label.

I  think the argument "k8s provides users with labels, kubelet should too" is a false equivalence. K8s is providing flexibility to users, a lower level interface between kubelet and a container runtime doesn't need that flexibility necessarily (unless you provide concrete examples), it just needs a way to filter by a limited set of items. And those items can be edited over time to ask runtimes to implement further filters (be it by labels or what have you).